extensions/trackedfile/src/main/java/io/deephaven/extensions/trackedfile/CompletableLocalOutputStream.java

```diff
@@ -0,0 +1,232 @@
+     */
+    @Nullable
+    private static File prepareDestinationFileLocation(@NotNull final File destFile) {
+        final File destination = destFile.getAbsoluteFile();
+        if (destination.exists()) {
+            if (destination.isDirectory()) {
+                throw new UncheckedDeephavenException(
+                        String.format("Destination %s exists and is a directory", destination));
+            }
+            if (!destination.canWrite()) {
+                throw new UncheckedDeephavenException(
+                        String.format("Destination %s exists but is not writable", destination));
+            }
+            return null;
+        }
+        final File firstParent = destination.getParentFile();
+        if (firstParent.isDirectory()) {
+            if (firstParent.canWrite()) {
+                return null;
+            }
+            throw new UncheckedDeephavenException(
+                    String.format("Destination %s has non writable parent directory", destination));
+        }
+        File firstCreated = firstParent;
+        File parent;
+        for (parent = destination.getParentFile(); parent != null && !parent.exists(); parent =
+                parent.getParentFile()) {
+            firstCreated = parent;
+        }
+        if (parent == null) {
+            throw new IllegalArgumentException(
+                    String.format("Can't find any existing parent directory for destination path: %s", destination));
+        }
+        if (!parent.isDirectory()) {
+            throw new IllegalArgumentException(
+                    String.format("Existing parent file %s of %s is not a directory", parent, destination));
+        }
+        if (!firstParent.mkdirs()) {
+            throw new UncheckedDeephavenException("Couldn't (re)create destination directory " + firstParent);
+        }
+        return firstCreated;
+    }
+
+
+    /**
+     * Backup any existing files at destination and rename the shadow file to destination file.
+     */
+    private static void installShadowFile(@NotNull final File destFile, @NotNull final File shadowDestFile) {
+        final File backupDestFile = getBackupFile(destFile);
+        if (destFile.exists() && !destFile.renameTo(backupDestFile)) {
+            throw new UncheckedDeephavenException(
+                    String.format("Failed to install shadow file at %s because a file already exists at the path " +
+                            "which " + "couldn't be renamed to %s", destFile.getAbsolutePath(),
+                            backupDestFile.getAbsolutePath()));
+        }
+        if (!shadowDestFile.exists()) {
+            throw new UncheckedDeephavenException(
+                    String.format("Failed to install shadow file at %s because shadow file doesn't exist at %s",
+                            destFile.getAbsolutePath(), shadowDestFile.getAbsolutePath()));
+        }
```

Comment:
On the one hand, this makes the error more clear. OTOH, it's doing a stat. I would skip it.